File: addTwoNumbers/addTwoNumbers.java

class ListNode {
    int val;
    ListNode next;
    ListNode() {}
    ListNode(int val) { this.val = val; }
    ListNode(int val, ListNode next) { this.val = val; this.next = next; }
}
 
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode sumNode = new ListNode();
        ListNode curr = sumNode;
        int carry = 0;

        while (l1 != null || l2 != null || carry != 0) {
            int l1Val = l1 != null ? l1.val : 0;
            int l2Val = l2 != null ? l2.val : 0;

            int columnVal = l1Val + l2Val + carry;

            carry = (int)(columnVal / 10);
            ListNode newNode = new ListNode(columnVal % 10);

            curr.next = newNode;
            curr = newNode;

            l1 = l1 != null ? l1.next : null;
            l2 = l2 != null ? l2.next : null;
        }

        return sumNode.next;
    }
}

File: addTwoNumbers/addTwoNumbers.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
    const sumNode = new ListNode();
    let curr = sumNode;
    let carry = 0;

    while (l1 != null || l2 != null || carry != 0) {
        const l1Val = l1 ? l1.val : 0;
        const l2Val = l2 ? l2.val : 0;

        const columnVal = l1Val + l2Val + carry;

        carry = Math.floor(columnVal / 10);
        const newNode = new ListNode(columnVal % 10);

        curr.next = newNode;
        curr = newNode;

        l1 = l1 ? l1.next : null;
        l2 = l2 ? l2.next : null;
    }

    return sumNode.next;
};